File: 2020/rust/src/day1.rs
use simple_error::SimpleError;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;

pub fn day1(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 1: Report Repair");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    println!("\tPart 1: {}", part1(&input)?);
    println!("\tPart 2: {}", part2(&input)?);
    Ok(())
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    let entries: Result<Vec<i32>, _> = input.lines().map(|x| x.parse::<i32>()).collect();
    match entries {
        Ok(entries) => {
            for i in 0..entries.len() {
                for j in i + 1..entries.len() {
                    if entries[i] + entries[j] == 2020 {
                        return Ok(entries[i] * entries[j]);
                    }
                }
            }
            Err(SimpleError::new("Solution not found").into())
        }
        Err(err) => Err(Box::new(err)),
    }
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    let entries: Result<Vec<i32>, _> = input.lines().map(|x| x.parse::<i32>()).collect();
    match entries {
        Ok(entries) => {
            for i in 0..entries.len() {
                for j in i + 1..entries.len() {
                    for k in j + 1..entries.len() {
                        if entries[i] + entries[j] + entries[k] == 2020 {
                            return Ok(entries[i] * entries[j] * entries[k]);
                        }
                    }
                }
            }
            Err(SimpleError::new("Solution not found").into())
        }
        Err(err) => Err(Box::new(err)),
    }
}

File: 2020/go/day1.go
package advent2020

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func Day1(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 1: Report Repair")
	part1, err := day1Part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %d\n", part1)

	part2, err := day1Part2(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %d\n", part2)

	return nil
}

func inputToIntSlice(input string) ([]int, error) {
	lines := strings.Split(input, "\n")
	entries := make([]int, len(lines))
	var err error
	for i, line := range lines {
		entries[i], err = strconv.Atoi(line)
		if err != nil {
			return nil, err
		}
	}
	return entries, nil
}

func day1Part1(input string) (int, error) {
	entries, err := inputToIntSlice(input)
	if err != nil {
		return 0, err
	}

	for i, x := range entries {
		for _, y := range entries[i+1:] {
			if x+y == 2020 {
				return x * y, nil
			}
		}
	}
	return 0, fmt.Errorf("Solution not found")
}

func day1Part2(input string) (int, error) {
	entries, err := inputToIntSlice(input)
	if err != nil {
		return 0, err
	}

	for i, x := range entries {
		for j, y := range entries[i+1:] {
			for _, z := range entries[j+1:] {
				if x+y+z == 2020 {
					return x * y * z, nil
				}
			}
		}
	}
	return 0, fmt.Errorf("Solution not found")
}
